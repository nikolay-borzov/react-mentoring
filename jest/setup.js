import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
