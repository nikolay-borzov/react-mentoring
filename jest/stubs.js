export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre'],
  vote_average: 7.5
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
