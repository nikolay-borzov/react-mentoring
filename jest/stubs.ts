import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
