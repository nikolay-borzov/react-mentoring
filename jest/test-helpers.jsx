import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
