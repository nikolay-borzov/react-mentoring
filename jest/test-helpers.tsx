import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
