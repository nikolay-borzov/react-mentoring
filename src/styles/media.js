// @ flow

import { css } from 'styled-components'

const sizes = {
  small: 768,
  medium: 992,
  large: 1200
}

export const media = {
  small: (...args) => css`
    @media (min-width: ${sizes.small}px) and (max-width: ${sizes.medium -
        1}px) {
      ${css(...args)};
    }
  `,

  medium: (...args) => css`
    @media (min-width: ${sizes.medium}px) and (max-width: ${sizes.large -
        1}px) {
      ${css(...args)};
    }
  `,

  large: (...args) => css`
    @media (min-width: ${sizes.large}px) {
      ${css(...args)};
    }
  `,

  bigger: (...args) => css`
    @media (min-width: ${sizes.small}px) {
      ${css(...args)};
    }
  `
}
