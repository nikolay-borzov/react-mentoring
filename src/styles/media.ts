import { css, CSSObject, SimpleInterpolation } from 'styled-components'

const sizes = {
  small: 768,
  medium: 992,
  large: 1200
}

type CSS = TemplateStringsArray | CSSObject

export const media = {
  small: (first: CSS, ...interpolations: SimpleInterpolation[]) => css`
    @media (min-width: ${sizes.small}px) and (max-width: ${sizes.medium -
        1}px) {
      ${css(first, ...interpolations)};
    }
  `,

  medium: (first: CSS, ...interpolations: SimpleInterpolation[]) => css`
    @media (min-width: ${sizes.medium}px) and (max-width: ${sizes.large -
        1}px) {
      ${css(first, ...interpolations)};
    }
  `,

  large: (first: CSS, ...interpolations: SimpleInterpolation[]) => css`
    @media (min-width: ${sizes.large}px) {
      ${css(first, ...interpolations)};
    }
  `,

  bigger: (first: CSS, ...interpolations: SimpleInterpolation[]) => css`
    @media (min-width: ${sizes.small}px) {
      ${css(first, ...interpolations)};
    }
  `
}
