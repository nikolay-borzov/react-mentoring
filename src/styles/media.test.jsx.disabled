import React from 'react'
import styled from 'styled-components'
import 'jest-styled-components'

import { itRendersCorrectlyShallow } from '../../jest/test-helpers'

import { media } from './media'

describe('media', () => {
  itRendersCorrectlyShallow(() => {
    const Component = styled.div`
      ${media.small`z-index: 2`};
    `

    return <Component />
  }, `sets styles for 'small' media`)

  itRendersCorrectlyShallow(() => {
    const Component = styled.div`
      ${media.medium`z-index: 2`};
    `

    return <Component />
  }, `sets styles for 'medium' media`)

  itRendersCorrectlyShallow(() => {
    const Component = styled.div`
      ${media.large`z-index: 2`};
    `

    return <Component />
  }, `sets styles for 'large' media`)

  itRendersCorrectlyShallow(() => {
    const Component = styled.div`
      ${media.bigger`z-index: 2`};
    `

    return <Component />
  }, `sets styles for 'bigger' media`)
})
