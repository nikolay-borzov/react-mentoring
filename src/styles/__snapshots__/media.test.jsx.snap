// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media sets styles for 'bigger' media 1`] = `
@media (min-width:768px) {
  .c0 {
    z-index: 2;
  }
}

<div
  className="c0"
/>
`;

exports[`media sets styles for 'large' media 1`] = `
@media (min-width:1200px) {
  .c0 {
    z-index: 2;
  }
}

<div
  className="c0"
/>
`;

exports[`media sets styles for 'medium' media 1`] = `
@media (min-width:992px) and (max-width:1199px) {
  .c0 {
    z-index: 2;
  }
}

<div
  className="c0"
/>
`;

exports[`media sets styles for 'small' media 1`] = `
@media (min-width:768px) and (max-width:991px) {
  .c0 {
    z-index: 2;
  }
}

<div
  className="c0"
/>
`;
